contracts/IFTBase.sol

```diff
@@ -0,0 +1,237 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import { IIFTMsgs } from "./msgs/IIFTMsgs.sol";
+import { IIBCAppCallbacks } from "./msgs/IIBCAppCallbacks.sol";
+import { IICS27GMPMsgs } from "./msgs/IICS27GMPMsgs.sol";
+
+import { IIFT } from "./interfaces/IIFT.sol";
+import { IICS27GMP } from "./interfaces/IICS27GMP.sol";
+import { IIFTSendCallConstructor } from "./interfaces/IIFTSendCallConstructor.sol";
+import { IIBCSenderCallbacks } from "./interfaces/IIBCSenderCallbacks.sol";
+import { IIFTErrors } from "./errors/IIFTErrors.sol";
+
+import { ERC20 } from "@openzeppelin-contracts/token/ERC20/ERC20.sol";
+import { AccessManaged } from "@openzeppelin-contracts/access/manager/AccessManaged.sol";
+import { IBCCallbackReceiver } from "./utils/IBCCallbackReceiver.sol";
+
+/// @title IFT Base Contract
+/// @notice Abstract base contract for Interchain Fungible Tokens
+/// @dev Extend this contract and implement the ERC20 constructor to create an IFT token
+abstract contract IFTBase is IIFTErrors, IIFT, ERC20, AccessManaged, IBCCallbackReceiver {
+    /// @notice Storage for IFT-specific state
+    /// @param _ics27Gmp The ICS27-GMP contract for sending cross-chain messages
+    /// @param _iftBridges Mapping of client IDs to their bridge configurations
+    /// @param _pendingTransfers Mapping of (clientId, sequence) to pending transfer info
+    struct IFTStorage {
+        IICS27GMP _ics27Gmp;
+        mapping(string clientId => IIFTMsgs.IFTBridge bridge) _iftBridges;
+        mapping(string clientId => mapping(uint64 seq => IIFTMsgs.PendingTransfer info)) _pendingTransfers;
+    }
+
+    /// @notice ERC-7201 slot for the IFT storage
+    /// @dev keccak256(abi.encode(uint256(keccak256("ibc.storage.IFT")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant IFT_STORAGE_SLOT = 0x35d0029e62ce5824ad5e38215107659b8aa50b0046e8bc44a0f4a32b87d61a00;
+
+    /// @notice Default timeout duration for IFT transfers (15 minutes)
+    uint64 private constant DEFAULT_TIMEOUT_DURATION = 15 minutes;
+
+    /// @notice Initializes the IFT base contract
+    /// @param ics27Gmp_ The ICS27-GMP contract address
+    /// @param authority_ The AccessManager contract address for access control
+    constructor(IICS27GMP ics27Gmp_, address authority_) AccessManaged(authority_) {
+        IFTStorage storage $ = _getIFTStorage();
+        $._ics27Gmp = ics27Gmp_;
+    }
```

Comment:
The storage slot implies this is an upgradable contract, so we probably should use an initializer instead of constructor.